### PR TITLE
[WIP] Add exploratory code for scripting scenarios

### PR DESCRIPTION
**Pull request by lambdam**
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

This is a PR for the bounty [#17824](https://github.com/status-im/status-mobile/issues/17824)

### Summary

[comment]: # (Summarise the problem and how the pull request solves it)

We try to script scenarios that can be run programmatically or from the REPL and that would bring the application in a desired state.

#### Platforms

- Android
- iOS
- ~~macOS~~
- ~~Linux~~
- ~~Windows~~

##### Functional

- 1-1 chats
- public chats
- group chats
- wallet / transactions
- ~~dapps / app browsing~~
- ~~account recovery~~
- new account
- user profile updates
- ~~networks~~
- ~~mailservers~~
- ~~fleet~~
- ~~bootnodes~~

##### Non-functional

- ~~battery performance~~
- ~~CPU performance / speed of the app~~
- ~~network consumption~~

### Steps to test

- Launch a fresh application with no profile created
- Open the `utils.scenario` namespace in your text editor and evaluate the code.
- Execute the `run-init-scenario!` function in the source file (in the comment block) or in the REPL.

status: wip

### CODE REVIEW

**Review by ilmotta — COMMENTED** [review]
@lambdam Thank you for submitting this WIP PR.

I think you nailed the overall idea to process and *listen* to events, though my overall analysis is that the solution is overly complicated.

I see the PR's code is reinventing the wheel to flatten async code using macros. We just need to embrace promises. It's a proven model. We have other parts of the `status-mobile` repository that avoid promises like the plague and use callbacks, but they are harder to reason and unnecessarily nested, and we'll refactor them one day. Promises can also help us handle errors in the scenarios very cleanly and flexibly because the developer can attach `catch` calls in the middle of any step.

I'm sharing a concrete example below. It works well to showcase how scenarios can be defined. It also timeout when the expected event is not triggered for the default 5s window.

```clojure
(ns dev.account-flows
  (:require
    [promesa.core :as p]
    [re-frame.core :as re-frame]
    [utils.re-frame :as rf]))

(defn sleep
  [delay-ms]
  (js/Promise.
   (fn [promise-resolve]
     (js/setTimeout promise-resolve delay-ms))))

(defn wait-for
  ([target-event]
   (wait-for target-event 5000))
  ([target-event timeout-ms]
   (js/Promise.
    (fn [promise-resolve promise-reject]
      (let [cb-id    (gensym "post-event-callback")
            timer-id (js/setTimeout (fn []
                                      (re-frame/remove-post-event-callback cb-id)
                                      (promise-reject ::timeout))
                                    timeout-ms)]
        (re-frame/add-post-event-callback
         cb-id
         (fn [event]
           (when (= event target-event)
             (re-frame/remove-post-event-callback cb-id)
             (js/clearTimeout timer-id)
             (promise-resolve)))))))))

(comment
  ;; Example without promesa.
  (-> (js/Promise.resolve (rf/dispatch [:open-modal :new-to-status]))
      (.then #(sleep 1000))
      (.then #(rf/dispatch [:onboarding-2/navigate-to-create-profile]))
      (.then #(wait-for [:navigate-to-within-stack [:create-profile :new-to-status]]))
      (.then #(tap> :ok))
      (.catch #(tap> {:error %}))))

(comment
  ;; Example with promesa.
  (-> (p/do (rf/dispatch [:open-modal :new-to-status])
            (sleep 1000)
            (rf/dispatch [:onboarding-2/navigate-to-create-profile])
            (wait-for [:navigate-to-within-stack [:create-profile :new-to-status]])
            (tap> :ok))
      (p/catch #(tap> {:error %}))))
```

Also worth sharing we should avoid mixing dev-only code with production code, so it's better to write it somewhere outside `utils`.

--------

In summary, I believe the final solution should be approached without macros, or at least without nesting callbacks (thus using promises). The example code I shared is a good starting point.

But to reiterate, what matters to solve #17824 is to achieve a reliable solution to define scenarios based on re-frame events to create accounts and make them mutual contacts.

**Comment by flexsurfer** [comment]
hey @J-Son89 still relevant?

**Comment by J-Son89** [comment]
Thanks @flexsurfer - will close for now.
